include/Types.hh

```diff
@@ -0,0 +1,187 @@
+        while (m_size > 0) {
+            pop_back();
+        }
+
+        operator delete(m_data);
+    }
+
+    /// @brief Copies a new element into the array.
+    /// @param obj The object to copy.
+    /// @return A reference to the object.
+    T &push_back(const T &obj) {
+        ASSERT(initialized() && !full());
+        new (m_data + m_size++) T(obj);
+        return back();
+    }
+
+    /// @brief Moves a new element into the array.
+    /// @param obj The object to move.
+    /// @return A reference to the object.
+    T &push_back(T &&obj) {
+        ASSERT(initialized() && !full());
+        new (m_data + m_size++) T(std::move(obj));
+        return back();
+    }
+
+    /// @brief Creates a new element in-place in the array.
+    /// @tparam ...Args Variadic template for packing.
+    /// @param ...args Arguments to the constructor.
+    /// @return A reference to the object.
+    template <typename... Args>
+    T &emplace_back(Args &&...args) {
+        ASSERT(initialized() && !full());
+        new (m_data + m_size++) T(std::forward<Args>(args)...);
+        return back();
+    }
+
+    /// @brief Deletes the last existing element from the array.
+    void pop_back() {
+        ASSERT(initialized() && !empty());
+        m_data[--m_size].~T();
+    }
+
+    /// @brief Initializes the vector with the provided capacity.
+    /// @param capacity The number of elements to initialize the vector with.
+    void reserve(size_t capacity) {
+        ASSERT(!initialized());
+        allocate(capacity);
+    }
+
+    /// @brief Checks if there are no existing elements in the array.
+    /// @return True if the array is empty, otherwise false.
+    bool empty() const {
+        return m_size == 0;
+    }
+
+    /// @brief Checks if all elements exist in the array.
+    /// @return True if the array is full, otherwise false.
+    bool full() const {
+        return m_size == m_capacity;
+    }
+
+    /// @brief Checks if the array exists and if the capacity is non-zero.
+    /// @return True if the array is initialized, otherwise false.
+    bool initialized() const {
+        return m_data && m_capacity != 0;
+    }
+
+    /// @brief Gets the first element in the array, whether it exists or not.
+    /// @return The first element in the array.
+    T &front() {
+        return *m_data;
+    }
+
+    /// @brief Gets the first element in the array, whether it exists or not.
+    /// @return The first element in the array.
+    const T &front() const {
+        return *m_data;
+    }
+
+    /// @brief Gets the last existing element in the array, or the first element if none exist.
+    /// @return The last existing element in the array.
+    T &back() {
+        return *(m_data + m_size);
+    }
+
+    /// @brief Gets the last existing element in the array, or the first element if none exist.
+    /// @return The last existing element in the array.
+    const T &back() const {
+        return *(m_data + m_size);
+    }
```

Comment:
I think you have an off-by-one error. `back()` should not return the same as `end()`
```suggestion
    T &back() {
        return m_size > 0 ? *(m_data + m_size - 1) : *m_data;
    }

    /// @brief Gets the last existing element in the array, or the first element if none exist.
    /// @return The last existing element in the array.
    const T &back() const {
        return m_size > 0 ? *(m_data + m_size - 1) : *m_data;
    }
```